# _Evolutionary analysis of genome-specific duplications in flatworm genomes_ (Langleib et _al_., 2024)

## Overview
This repository contains data associated with the manuscript "_Evolutionary analysis of genome-specific duplications in flatworm genomes_" (Langleib et _al_., 2024). All necessary data needed to perform the analyses described in the text can be found here, including the raw data and the resultant files (up to 100 MB, due to GitHub storage limits).

Further data is available on Zenodo: **https://zenodo.org/doi/10.5281/zenodo.10257624**.

## Main directories and their content
This repository is structured with three main directories:

- **data**: This directory contains the raw data necessary for the study.
- **notebooks**: This directory contains notebooks with the code necessary to produce every table or figure of the study, as well as associated scripts. Each notebook includes an overview of its main task.
- **results**: This directory contains all the intermediate and final files produced after running the notebooks on the data.

___
**Important note**

Most of the data produced utilizes an **internal code for genes** (explicitly generated in the second notebook provided) to prevent crashes from various programs used (_e.g._, get_homologues). This internal code is **available at `results/misc/gene_code_correspondence.tsv`**.

Additionally, a table with **all inparalogous groups** detected in this study is **available at `results/misc/inparalog_code_correspondence.tsv`**.
___

## Any questions? Please contact us!
If you have questions regarding this study or the associated data, please don't hesitate to contact Msc. Mauricio Langleib at mauricio.langleib@gmail.com, mlangleib@higiene.edu.uy, or mlangleib@pasteur.edu.uy, or Dr. Andr√©s Iriarte at airiarteo@higiene.edu.uy.
