{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Overview\n",
    "In this notebook tables presented in this paper are either created or referenced (in case they were created in a previous notebook, during analysis of data)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preliminary work: classify analyzed proteins\n",
    "Many of the following tables use information for each protein analyzed in this paper and the classifications performed. Recapitulating this into a python dictionary in order to make table creation easier."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import glob\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "# create a dictionary of proteins for each species\n",
    "# get empty dictionaries for all proteins of each species\n",
    "\n",
    "fasta_files = glob.glob('../data/platyhelminthes_dataset/*')\n",
    "# Dictionary to hold everything\n",
    "species_dict = {}\n",
    "# Define the categories for the innermost dictionary\n",
    "categories = [\"SignalP\", \"SecretomeP\", \"TMHMM\", \"TargetP\", \"Secretome\", \"Transmembranome\", \"Homologs\", \"Singleton\", \"Inparalog\"]\n",
    "# Process each FASTA file\n",
    "for fasta_file in fasta_files:\n",
    "    # Extract species name from the file path\n",
    "    parts = fasta_file.split('.')\n",
    "    species_name = parts[-6].rpartition('/')[2].split('_')\n",
    "    species = f\"{species_name[0].capitalize()[0]}. {species_name[1]}\"\n",
    "    \n",
    "    # Initialize the species dictionary if not already done\n",
    "    if species not in species_dict:\n",
    "        species_dict[species] = {}\n",
    "    \n",
    "    # Parse the FASTA file\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "        # Each record ID becomes a key in the species dictionary\n",
    "        species_dict[species][record.id] = {key: {} for key in categories}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "new2old_dict = {row['Original Name']: row['New Name'] for index,row in pd.read_csv('../results/misc/gene_code_correspondance.tsv', sep = '\\t').iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "secretome_and_transmembranome_predictions = pd.read_csv('../results/secretome_and_transmembranome_predictions/secretome_and_transmembranome_predictions.tsv', sep ='\\t' ).applymap(lambda x: new2old_dict.get(x, x))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SecretomeP</th>\n",
       "      <th>SignalP</th>\n",
       "      <th>TargetP</th>\n",
       "      <th>TMHMM</th>\n",
       "      <th>Secretome</th>\n",
       "      <th>Transmembranome</th>\n",
       "      <th>Mitochondrial</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Smed.18673</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Smed.18733</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Smed.1254</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Smed.21065</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Smed.11902</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           ID SecretomeP SignalP TargetP TMHMM Secretome Transmembranome  \\\n",
       "0  Smed.18673          Y       N       N     N         Y               N   \n",
       "1  Smed.18733          Y       N       N     N         Y               N   \n",
       "2   Smed.1254          Y       N       N     N         Y               N   \n",
       "3  Smed.21065          Y       N       N     N         Y               N   \n",
       "4  Smed.11902          Y       N       N     N         Y               N   \n",
       "\n",
       "  Mitochondrial  \n",
       "0             N  \n",
       "1             N  \n",
       "2             N  \n",
       "3             N  \n",
       "4             N  "
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "secretome_and_transmembranome_predictions.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['T. solium',\n",
       " 'M. lignano',\n",
       " 'T. regenti',\n",
       " 'H. microstoma',\n",
       " 'T. saginata',\n",
       " 'H. diminuta',\n",
       " 'F. hepatica',\n",
       " 'S. mattheei',\n",
       " 'S. curassoni',\n",
       " 'E. multilocularis',\n",
       " 'S. mansoni',\n",
       " 'T. asiatica',\n",
       " 'S. margrebowiei',\n",
       " 'C. sinensis',\n",
       " 'E. granulosus',\n",
       " 'M. corti',\n",
       " 'S. haematobium',\n",
       " 'S. japonicum',\n",
       " 'E. canadensis',\n",
       " 'S. mediterranea',\n",
       " 'O. viverrini']"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "[x.split('/')[-1].split('.')[-6].split('_')[0].capitalize()[0]+'. '+x.split('/')[-1].split('.')[-6].split('_')[1] for x \n",
    " in glob.glob('../data/platyhelminthes_dataset/*')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get the SignalP, SecretomeP, TMHMM, TargetP and classification into Secretome, Transmembranome or neither for each protein\n",
    "\n",
    "import pandas as pd\n",
    "import glob\n",
    "from Bio import SeqIO\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "\n",
    "# Assuming FASTA files are structured and named in a specific way that allows species identification\n",
    "fasta_files = glob.glob('../data/platyhelminthes_dataset/*')\n",
    "\n",
    "# Function to extract species name from FASTA file path\n",
    "def get_species_from_fasta_filename(filename):\n",
    "    species_name = filename.split('/')[-1].split('.')[-6].split('_')[0].capitalize()[0]+'. '+filename.split('/')[-1].split('.')[-6].split('_')[1]\n",
    "    return species_name\n",
    "\n",
    "# Dictionary to hold species, and their protein information\n",
    "species_dict = {}\n",
    "for fasta_file in fasta_files:\n",
    "    species = get_species_from_fasta_filename(fasta_file)\n",
    "    if species not in species_dict:\n",
    "        species_dict[species] = {}\n",
    "    for record in SeqIO.parse(fasta_file, 'fasta'):\n",
    "        species_dict[species][record.id] = {key: {} for key in [\"SignalP\", \"SecretomeP\", \"TMHMM\", \"TargetP\", \"Secretome\", \"Transmembranome\", \"Homologs\", \"Singleton\", \"Inparalog\"]}\n",
    "\n",
    "def get_species_from_protein_id(protein_id):\n",
    "    # Search for the species in the dictionary by protein ID\n",
    "    for species, proteins in species_dict.items():\n",
    "        if protein_id in proteins:\n",
    "            return species\n",
    "    return None\n",
    "\n",
    "def update_protein_info(row):\n",
    "    protein_id = row['ID']\n",
    "    species = get_species_from_protein_id(protein_id)\n",
    "    if species:\n",
    "        for category in ['SecretomeP', 'SignalP', 'TargetP', 'TMHMM', 'Secretome', 'Transmembranome', 'Mitochondrial']:  # Skip the 'ID' column\n",
    "            species_dict[species][protein_id][category] = row[category]\n",
    "\n",
    "# Use ThreadPoolExecutor to update the dictionary\n",
    "with ThreadPoolExecutor(max_workers=32) as executor:\n",
    "    executor.map(update_protein_info, [row for index, row in secretome_and_transmembranome_predictions.iterrows()])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>SecretomeP</th>\n",
       "      <th>SignalP</th>\n",
       "      <th>TargetP</th>\n",
       "      <th>TMHMM</th>\n",
       "      <th>Secretome</th>\n",
       "      <th>Transmembranome</th>\n",
       "      <th>Mitochondrial</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>285305</th>\n",
       "      <td>Tsol.1333</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285310</th>\n",
       "      <td>Tsol.1894</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285311</th>\n",
       "      <td>Tsol.11358</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285314</th>\n",
       "      <td>Tsol.10360</td>\n",
       "      <td>Y</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>285318</th>\n",
       "      <td>Tsol.12227</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297546</th>\n",
       "      <td>Tsol.12133</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297547</th>\n",
       "      <td>Tsol.10652</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297558</th>\n",
       "      <td>Tsol.11696</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297559</th>\n",
       "      <td>Tsol.1392</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>297562</th>\n",
       "      <td>Tsol.11157</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "      <td>Y</td>\n",
       "      <td>N</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3446 rows × 8 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                ID SecretomeP SignalP TargetP TMHMM Secretome Transmembranome  \\\n",
       "285305   Tsol.1333          Y       N       N     N         Y               N   \n",
       "285310   Tsol.1894          Y       N       N     N         Y               N   \n",
       "285311  Tsol.11358          Y       N       N     Y         N               Y   \n",
       "285314  Tsol.10360          Y       Y       N     N         Y               N   \n",
       "285318  Tsol.12227          Y       N       N     Y         N               Y   \n",
       "...            ...        ...     ...     ...   ...       ...             ...   \n",
       "297546  Tsol.12133          N       N       N     Y         N               Y   \n",
       "297547  Tsol.10652          N       Y       N     Y         N               Y   \n",
       "297558  Tsol.11696          N       N       N     N         N               N   \n",
       "297559   Tsol.1392          N       N       N     N         N               N   \n",
       "297562  Tsol.11157          N       N       N     Y         N               Y   \n",
       "\n",
       "       Mitochondrial  \n",
       "285305             N  \n",
       "285310             N  \n",
       "285311             N  \n",
       "285314             N  \n",
       "285318             N  \n",
       "...              ...  \n",
       "297546             N  \n",
       "297547             N  \n",
       "297558             N  \n",
       "297559             N  \n",
       "297562             N  \n",
       "\n",
       "[3446 rows x 8 columns]"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "secretome_and_transmembranome_predictions.query(\"ID.str.contains('Tsol.1')\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# get status as homolog or singleton, as well as inparalog status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "inparalogs_list = [new2old_dict.get(gene) for genes in pd.read_csv('../results/misc/inparalogs_group_composition_final.tsv', sep ='\\t').Genes.to_list() for gene in genes.split(', ')]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 21/21 [09:35<00:00, 27.40s/it]\n"
     ]
    }
   ],
   "source": [
    "# Path to the dataset of homolog groups\n",
    "homolog_files = glob.glob('../results/homolog_groups/*')\n",
    "# Dictionary to track group size and proteins\n",
    "group_info = {}\n",
    "# Populate group information\n",
    "for file in homolog_files:\n",
    "    proteins = list(SeqIO.parse(file, 'fasta'))\n",
    "    group_size = len(proteins)\n",
    "    for protein in proteins:\n",
    "        group_info[protein.id] = {'size': group_size, 'group_file': file}\n",
    "# Update species_dict based on homolog group size and inparalog status\n",
    "for species, proteins in tqdm.tqdm(species_dict.items()):\n",
    "    for protein_id in proteins:\n",
    "        if protein_id in group_info:\n",
    "            if group_info[protein_id]['size'] == 1:\n",
    "                species_dict[species][protein_id]['Homologs'] = 'N'\n",
    "                species_dict[species][protein_id]['Singleton'] = 'Y'\n",
    "            else:\n",
    "                species_dict[species][protein_id]['Homologs'] = 'Y'\n",
    "                species_dict[species][protein_id]['Singleton'] = 'N'\n",
    "        # Inparalog status update\n",
    "        species_dict[species][protein_id]['Inparalog'] = 'Y' if protein_id in inparalogs_list else 'N'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Creating a dictionary with correspondance between species and genome codes (needed to retrieve assembly data from WormBase ParaSite)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "code2assembly = {'CSIN':  'CSKR.v2',\n",
    "                 'ECAN':  'ECANG7',\n",
    "                 'EGRA':  'EGRAN001',\n",
    "                 'EMUL':  'EMULTI002',\n",
    "                 'FHEP':  'Fasciola_10x_pilon',\n",
    "                 'HDIM':  'H.diminuta_WMSil1',\n",
    "                 'HMIC':  'HMN_v3',\n",
    "                 'MLIG':  'Mlig_3_7',\n",
    "                 'MCOR':  'M_corti_Specht_Voge_0011_upd',\n",
    "                 'OVIV':  'OpiViv1.0',\n",
    "                 'SCUR':  'tdSchCurr1.1',\n",
    "                 'SHAE':  'UoM_Shae.V3',\n",
    "                 'SJAP':  'ASM636876v1',\n",
    "                 'SMAN':  'SM_V10',\n",
    "                 'SMAR':  'tdSchMarg1.1',\n",
    "                 'SMAT':  'tdSchMatt1.1',\n",
    "                 'SMED':  'SmedGD_c1.3',\n",
    "                 'TASI':  'Taenia_asiatica_TASYD01_v1',\n",
    "                 'TSAG':  'ASM169307v2',\n",
    "                 'TSOL':  'Tsolium_Mexico_v1',\n",
    "                 'TREG':  'tdTriRege1.1'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "assemblies = list(code2assembly.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "platy_data = pd.read_csv('../data/species_Platyhelminthes--___.csv').query(\"Assembly in @assemblies\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(21, 10)"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "platy_data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now some InterPro annotations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "interpro_header = ['Protein accession',\n",
    "                   'Sequence MD5 digest',\n",
    "                   'Sequence length',\n",
    "                   'Analysis',\n",
    "                   'Signature accession', \n",
    "                   'Signature description', \n",
    "                   'Start location',\n",
    "                   'Stop location',\n",
    "                   'Score',\n",
    "                   'Status',\n",
    "                   'Date', \n",
    "                   'InterPro annotations - accession',\n",
    "                   'InterPro annotations - description', \n",
    "                   'GO annotations',\n",
    "                   'Pathways annotations']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "interpro_table = pd.concat([pd.read_csv(tsv_file, names = interpro_header, sep = '\\t') for tsv_file in glob.glob('../results/functional_mapping/interpro_annotation/*')])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Protein accession</th>\n",
       "      <th>Sequence MD5 digest</th>\n",
       "      <th>Sequence length</th>\n",
       "      <th>Analysis</th>\n",
       "      <th>Signature accession</th>\n",
       "      <th>Signature description</th>\n",
       "      <th>Start location</th>\n",
       "      <th>Stop location</th>\n",
       "      <th>Score</th>\n",
       "      <th>Status</th>\n",
       "      <th>Date</th>\n",
       "      <th>InterPro annotations - accession</th>\n",
       "      <th>InterPro annotations - description</th>\n",
       "      <th>GO annotations</th>\n",
       "      <th>Pathways annotations</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>maker-E.canG7_contigs_5369-pred_gff_augustus_m...</td>\n",
       "      <td>8783dc2c9fe014f0bd2933b3bba33279</td>\n",
       "      <td>115</td>\n",
       "      <td>Pfam</td>\n",
       "      <td>PF00102</td>\n",
       "      <td>Protein-tyrosine phosphatase</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>1.6E-19</td>\n",
       "      <td>T</td>\n",
       "      <td>15-04-2024</td>\n",
       "      <td>IPR000242</td>\n",
       "      <td>Tyrosine-specific protein phosphatase, PTPase ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>maker-E.canG7_contigs_5369-pred_gff_augustus_m...</td>\n",
       "      <td>8783dc2c9fe014f0bd2933b3bba33279</td>\n",
       "      <td>115</td>\n",
       "      <td>SUPERFAMILY</td>\n",
       "      <td>SSF52799</td>\n",
       "      <td>(Phosphotyrosine protein) phosphatases II</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>3.68E-25</td>\n",
       "      <td>T</td>\n",
       "      <td>15-04-2024</td>\n",
       "      <td>IPR029021</td>\n",
       "      <td>Protein-tyrosine phosphatase-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>maker-E.canG7_contigs_5369-pred_gff_augustus_m...</td>\n",
       "      <td>8783dc2c9fe014f0bd2933b3bba33279</td>\n",
       "      <td>115</td>\n",
       "      <td>Gene3D</td>\n",
       "      <td>G3DSA:3.90.190.10</td>\n",
       "      <td>Protein tyrosine phosphatase superfamily</td>\n",
       "      <td>1</td>\n",
       "      <td>114</td>\n",
       "      <td>1.1E-27</td>\n",
       "      <td>T</td>\n",
       "      <td>15-04-2024</td>\n",
       "      <td>IPR029021</td>\n",
       "      <td>Protein-tyrosine phosphatase-like</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>maker-E.canG7_contigs_5369-pred_gff_augustus_m...</td>\n",
       "      <td>8783dc2c9fe014f0bd2933b3bba33279</td>\n",
       "      <td>115</td>\n",
       "      <td>PANTHER</td>\n",
       "      <td>PTHR19134</td>\n",
       "      <td>RECEPTOR-TYPE TYROSINE-PROTEIN PHOSPHATASE</td>\n",
       "      <td>1</td>\n",
       "      <td>113</td>\n",
       "      <td>1.8E-21</td>\n",
       "      <td>T</td>\n",
       "      <td>15-04-2024</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>maker-E.canG7_contigs_5369-pred_gff_augustus_m...</td>\n",
       "      <td>8783dc2c9fe014f0bd2933b3bba33279</td>\n",
       "      <td>115</td>\n",
       "      <td>ProSiteProfiles</td>\n",
       "      <td>PS50055</td>\n",
       "      <td>PTP type protein phosphatase domain profile.</td>\n",
       "      <td>1</td>\n",
       "      <td>115</td>\n",
       "      <td>14.818786</td>\n",
       "      <td>T</td>\n",
       "      <td>15-04-2024</td>\n",
       "      <td>IPR000242</td>\n",
       "      <td>Tyrosine-specific protein phosphatase, PTPase ...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   Protein accession  \\\n",
       "0  maker-E.canG7_contigs_5369-pred_gff_augustus_m...   \n",
       "1  maker-E.canG7_contigs_5369-pred_gff_augustus_m...   \n",
       "2  maker-E.canG7_contigs_5369-pred_gff_augustus_m...   \n",
       "3  maker-E.canG7_contigs_5369-pred_gff_augustus_m...   \n",
       "4  maker-E.canG7_contigs_5369-pred_gff_augustus_m...   \n",
       "\n",
       "                Sequence MD5 digest  Sequence length         Analysis  \\\n",
       "0  8783dc2c9fe014f0bd2933b3bba33279              115             Pfam   \n",
       "1  8783dc2c9fe014f0bd2933b3bba33279              115      SUPERFAMILY   \n",
       "2  8783dc2c9fe014f0bd2933b3bba33279              115           Gene3D   \n",
       "3  8783dc2c9fe014f0bd2933b3bba33279              115          PANTHER   \n",
       "4  8783dc2c9fe014f0bd2933b3bba33279              115  ProSiteProfiles   \n",
       "\n",
       "  Signature accession                         Signature description  \\\n",
       "0             PF00102                  Protein-tyrosine phosphatase   \n",
       "1            SSF52799     (Phosphotyrosine protein) phosphatases II   \n",
       "2   G3DSA:3.90.190.10      Protein tyrosine phosphatase superfamily   \n",
       "3           PTHR19134    RECEPTOR-TYPE TYROSINE-PROTEIN PHOSPHATASE   \n",
       "4             PS50055  PTP type protein phosphatase domain profile.   \n",
       "\n",
       "   Start location  Stop location      Score Status        Date  \\\n",
       "0               1            113    1.6E-19      T  15-04-2024   \n",
       "1               1            113   3.68E-25      T  15-04-2024   \n",
       "2               1            114    1.1E-27      T  15-04-2024   \n",
       "3               1            113    1.8E-21      T  15-04-2024   \n",
       "4               1            115  14.818786      T  15-04-2024   \n",
       "\n",
       "  InterPro annotations - accession  \\\n",
       "0                        IPR000242   \n",
       "1                        IPR029021   \n",
       "2                        IPR029021   \n",
       "3                                -   \n",
       "4                        IPR000242   \n",
       "\n",
       "                  InterPro annotations - description  GO annotations  \\\n",
       "0  Tyrosine-specific protein phosphatase, PTPase ...             NaN   \n",
       "1                  Protein-tyrosine phosphatase-like             NaN   \n",
       "2                  Protein-tyrosine phosphatase-like             NaN   \n",
       "3                                                  -             NaN   \n",
       "4  Tyrosine-specific protein phosphatase, PTPase ...             NaN   \n",
       "\n",
       "   Pathways annotations  \n",
       "0                   NaN  \n",
       "1                   NaN  \n",
       "2                   NaN  \n",
       "3                   NaN  \n",
       "4                   NaN  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interpro_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "605473it [00:49, 12199.91it/s]\n"
     ]
    }
   ],
   "source": [
    "code2interpro = {row['Protein accession']: row['InterPro annotations - accession']+': '+row['InterPro annotations - description'] for index,row in tqdm.tqdm(interpro_table.iterrows())}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Table 1: Number of inparalogs and singletons identified in each species"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "if not os.path.exists('../results/tables'):\n",
    "    os.mkdir('../results/tables')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Assuming species_dict is structured as {species: {protein_id: {categories...}}}\n",
    "summary_data = []\n",
    "\n",
    "for species, proteins in species_dict.items():\n",
    "    total_proteins = len(proteins)\n",
    "    homologs_count = sum(1 for details in proteins.values() if details.get('Homologs') == 'Y')\n",
    "    singletons_count = sum(1 for details in proteins.values() if details.get('Singleton') == 'Y')\n",
    "    inparalogs_count = sum(1 for details in proteins.values() if details.get('Inparalog') == 'Y')\n",
    "    \n",
    "    summary_data.append({\n",
    "        'Species': species,\n",
    "        'Total Proteins': total_proteins,\n",
    "        'Homologs': homologs_count,\n",
    "        'Singletons': singletons_count,\n",
    "        'Inparalogs': inparalogs_count\n",
    "    })\n",
    "\n",
    "table_one = pd.DataFrame(summary_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "table_one.to_csv('../results/tables/table_1.tsv', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Total Proteins</th>\n",
       "      <th>Homologs</th>\n",
       "      <th>Singletons</th>\n",
       "      <th>Inparalogs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T. solium</td>\n",
       "      <td>12356</td>\n",
       "      <td>7741</td>\n",
       "      <td>4615</td>\n",
       "      <td>154</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M. lignano</td>\n",
       "      <td>49013</td>\n",
       "      <td>41292</td>\n",
       "      <td>7721</td>\n",
       "      <td>38828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T. regenti</td>\n",
       "      <td>14478</td>\n",
       "      <td>10466</td>\n",
       "      <td>4012</td>\n",
       "      <td>3259</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H. microstoma</td>\n",
       "      <td>10139</td>\n",
       "      <td>8120</td>\n",
       "      <td>2019</td>\n",
       "      <td>1004</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T. saginata</td>\n",
       "      <td>13160</td>\n",
       "      <td>10842</td>\n",
       "      <td>2318</td>\n",
       "      <td>673</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H. diminuta</td>\n",
       "      <td>15165</td>\n",
       "      <td>9937</td>\n",
       "      <td>5228</td>\n",
       "      <td>3023</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>F. hepatica</td>\n",
       "      <td>9731</td>\n",
       "      <td>4666</td>\n",
       "      <td>5065</td>\n",
       "      <td>247</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S. mattheei</td>\n",
       "      <td>10770</td>\n",
       "      <td>8108</td>\n",
       "      <td>2662</td>\n",
       "      <td>478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S. curassoni</td>\n",
       "      <td>10084</td>\n",
       "      <td>8804</td>\n",
       "      <td>1280</td>\n",
       "      <td>587</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>E. multilocularis</td>\n",
       "      <td>10663</td>\n",
       "      <td>9538</td>\n",
       "      <td>1125</td>\n",
       "      <td>515</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S. mansoni</td>\n",
       "      <td>9896</td>\n",
       "      <td>8844</td>\n",
       "      <td>1052</td>\n",
       "      <td>597</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>T. asiatica</td>\n",
       "      <td>13322</td>\n",
       "      <td>11188</td>\n",
       "      <td>2134</td>\n",
       "      <td>974</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>S. margrebowiei</td>\n",
       "      <td>9780</td>\n",
       "      <td>8651</td>\n",
       "      <td>1129</td>\n",
       "      <td>467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>C. sinensis</td>\n",
       "      <td>13489</td>\n",
       "      <td>7385</td>\n",
       "      <td>6104</td>\n",
       "      <td>297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>E. granulosus</td>\n",
       "      <td>10236</td>\n",
       "      <td>9243</td>\n",
       "      <td>993</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>M. corti</td>\n",
       "      <td>14704</td>\n",
       "      <td>8391</td>\n",
       "      <td>6313</td>\n",
       "      <td>2069</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S. haematobium</td>\n",
       "      <td>9431</td>\n",
       "      <td>7950</td>\n",
       "      <td>1481</td>\n",
       "      <td>343</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>S. japonicum</td>\n",
       "      <td>10089</td>\n",
       "      <td>7515</td>\n",
       "      <td>2574</td>\n",
       "      <td>389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>E. canadensis</td>\n",
       "      <td>11425</td>\n",
       "      <td>7300</td>\n",
       "      <td>4125</td>\n",
       "      <td>497</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>S. mediterranea</td>\n",
       "      <td>29850</td>\n",
       "      <td>11256</td>\n",
       "      <td>18594</td>\n",
       "      <td>9225</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>O. viverrini</td>\n",
       "      <td>16379</td>\n",
       "      <td>6508</td>\n",
       "      <td>9871</td>\n",
       "      <td>285</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Species  Total Proteins  Homologs  Singletons  Inparalogs\n",
       "0           T. solium           12356      7741        4615         154\n",
       "1          M. lignano           49013     41292        7721       38828\n",
       "2          T. regenti           14478     10466        4012        3259\n",
       "3       H. microstoma           10139      8120        2019        1004\n",
       "4         T. saginata           13160     10842        2318         673\n",
       "5         H. diminuta           15165      9937        5228        3023\n",
       "6         F. hepatica            9731      4666        5065         247\n",
       "7         S. mattheei           10770      8108        2662         478\n",
       "8        S. curassoni           10084      8804        1280         587\n",
       "9   E. multilocularis           10663      9538        1125         515\n",
       "10         S. mansoni            9896      8844        1052         597\n",
       "11        T. asiatica           13322     11188        2134         974\n",
       "12    S. margrebowiei            9780      8651        1129         467\n",
       "13        C. sinensis           13489      7385        6104         297\n",
       "14      E. granulosus           10236      9243         993         133\n",
       "15           M. corti           14704      8391        6313        2069\n",
       "16     S. haematobium            9431      7950        1481         343\n",
       "17       S. japonicum           10089      7515        2574         389\n",
       "18      E. canadensis           11425      7300        4125         497\n",
       "19    S. mediterranea           29850     11256       18594        9225\n",
       "20       O. viverrini           16379      6508        9871         285"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "table_one"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplementary Table 1\n",
    "This table was created manually, parsing data from WormBase ParaSite."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplementary Table 2: Summary of results for secreted and transmembrane protein-coding genes in each species analyzed. Mollusca and Annelida phyla were included."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First getting species2n50 and species2ncontig dictionaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dictionaries\n",
    "species2n50 = {}\n",
    "species2ncontig = {}\n",
    "\n",
    "for index,entry in platy_data.iterrows():\n",
    "    genus, species = entry[\"Species Name\"].split(' ')\n",
    "    key = f\"{genus[0]}. {species}\"\n",
    "    species2n50[key] = int(entry[\"N50\"].replace(\",\", \"\"))\n",
    "    species2ncontig[key] = entry[\"Number of Scaffolds\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the table with column names Genes and Annot\n",
    "file_path = '/home/mauricio/mauricio_PROYECTOS/PEIPER_PLATYS/platyhelminthes_inparalogs/results/functional_mapping/platyhelminthes_dataset.longest_isoforms.all.all.diamondCOGs.emapper.goatools_format.annotations'\n",
    "annotations_df = pd.read_csv(file_path, sep='\\t', names=['Genes', 'Annot'])\n",
    "\n",
    "# Update species_dict based on annotations\n",
    "for index, row in annotations_df.iterrows():\n",
    "    gene_id = row['Genes']\n",
    "    go_annotation = 'Y' if pd.notna(row['Annot']) and row['Annot'].strip() != '' else 'N'\n",
    "    \n",
    "    # Assuming gene_id is directly accessible as a key in species_dict\n",
    "    for species, proteins in species_dict.items():\n",
    "        if gene_id in proteins:\n",
    "            proteins[gene_id]['GOAnnotation'] = go_annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Initialize an empty list to store summary data for each species\n",
    "summary_data = []\n",
    "\n",
    "# Iterate through each species in the dictionary\n",
    "for species, proteins in species_dict.items():\n",
    "    total_proteins = len(proteins)\n",
    "    secreted_genes = sum(1 for details in proteins.values() if details.get('Secretome') == 'Y')\n",
    "    transmembrane_genes = sum(1 for details in proteins.values() if details.get('Transmembranome') == 'Y')\n",
    "    go_annotated_genes = sum(1 for details in proteins.values() if details.get('GOAnnotation') == 'Y')\n",
    "    \n",
    "    # Count the total number of Homologs and Inparalogs\n",
    "    total_homologs = sum(1 for details in proteins.values() if details.get('Homologs') == 'Y')\n",
    "    total_inparalogs = sum(1 for details in proteins.values() if details.get('Inparalog') == 'Y')\n",
    "    \n",
    "    # Count those Homologs and Inparalogs that also have GO annotations\n",
    "    homologs_with_go = sum(1 for details in proteins.values() if details.get('Homologs') == 'Y' and details.get('GOAnnotation') == 'Y')\n",
    "    inparalogs_with_go = sum(1 for details in proteins.values() if details.get('Inparalog') == 'Y' and details.get('GOAnnotation') == 'Y')\n",
    "    unannotated_inparalogs = sum(1 for details in proteins.values() if details.get('Inparalog') == 'Y' and details.get('GOAnnotation') == 'N')\n",
    "\n",
    "    # Calculating percentages\n",
    "    secreted_percent = (secreted_genes / total_proteins * 100) if total_proteins > 0 else 0\n",
    "    transmembrane_percent = (transmembrane_genes / total_proteins * 100) if total_proteins > 0 else 0\n",
    "    go_annotated_percent = (go_annotated_genes / total_proteins * 100) if total_proteins > 0 else 0\n",
    "    homologs_with_go_percent = (homologs_with_go / total_homologs * 100) if total_homologs > 0 else 0\n",
    "    inparalogs_with_go_percent = (inparalogs_with_go / total_inparalogs * 100) if total_inparalogs > 0 else 0\n",
    "    unannotated_inparalogs_percent = (unannotated_inparalogs / total_inparalogs * 100) if total_inparalogs > 0 else 0\n",
    "\n",
    "    # Append data for this species to the list\n",
    "    summary_data.append({\n",
    "        'Species': species,\n",
    "        'Total number of scaffolds': species2ncontig.get(species),\n",
    "        'N50': species2n50.get(species),\n",
    "        'Total number of proteins': total_proteins,\n",
    "        'Secreted protein-coding genes (%)': secreted_percent,\n",
    "        'Transmembrane protein-coding genes (%)': transmembrane_percent,\n",
    "        'Genes with GO annotation (%)': go_annotated_percent,\n",
    "        'Homologs with GO annotation (%)': homologs_with_go_percent,\n",
    "        'Inparalogs with GO annotation (%)': inparalogs_with_go_percent,\n",
    "        'Unannotated inparalogs (%)': unannotated_inparalogs_percent\n",
    "    })\n",
    "\n",
    "# Create a DataFrame from the summary data\n",
    "supp_table_two = pd.DataFrame(summary_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "supp_table_two.to_csv('../results/tables/supplementary_table_2.tsv', sep ='\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Species</th>\n",
       "      <th>Total number of scaffolds</th>\n",
       "      <th>N50</th>\n",
       "      <th>Total number of proteins</th>\n",
       "      <th>Secreted protein-coding genes (%)</th>\n",
       "      <th>Transmembrane protein-coding genes (%)</th>\n",
       "      <th>Genes with GO annotation (%)</th>\n",
       "      <th>Homologs with GO annotation (%)</th>\n",
       "      <th>Inparalogs with GO annotation (%)</th>\n",
       "      <th>Unannotated inparalogs (%)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>T. solium</td>\n",
       "      <td>11,237</td>\n",
       "      <td>67829</td>\n",
       "      <td>12356</td>\n",
       "      <td>7.389123</td>\n",
       "      <td>17.651343</td>\n",
       "      <td>71.819359</td>\n",
       "      <td>80.299703</td>\n",
       "      <td>74.025974</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>M. lignano</td>\n",
       "      <td>5,269</td>\n",
       "      <td>244885</td>\n",
       "      <td>49013</td>\n",
       "      <td>8.387571</td>\n",
       "      <td>22.573603</td>\n",
       "      <td>71.072981</td>\n",
       "      <td>73.827860</td>\n",
       "      <td>73.362007</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T. regenti</td>\n",
       "      <td>573</td>\n",
       "      <td>125840967</td>\n",
       "      <td>14478</td>\n",
       "      <td>8.281531</td>\n",
       "      <td>20.251416</td>\n",
       "      <td>69.836994</td>\n",
       "      <td>78.081406</td>\n",
       "      <td>62.964099</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>H. microstoma</td>\n",
       "      <td>7</td>\n",
       "      <td>25815276</td>\n",
       "      <td>10139</td>\n",
       "      <td>6.973074</td>\n",
       "      <td>19.666634</td>\n",
       "      <td>71.841404</td>\n",
       "      <td>78.522167</td>\n",
       "      <td>55.179283</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T. saginata</td>\n",
       "      <td>3,626</td>\n",
       "      <td>585232</td>\n",
       "      <td>13160</td>\n",
       "      <td>10.471125</td>\n",
       "      <td>19.247720</td>\n",
       "      <td>58.168693</td>\n",
       "      <td>62.506918</td>\n",
       "      <td>42.793462</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>H. diminuta</td>\n",
       "      <td>719</td>\n",
       "      <td>2331359</td>\n",
       "      <td>15165</td>\n",
       "      <td>8.578965</td>\n",
       "      <td>15.522585</td>\n",
       "      <td>55.351137</td>\n",
       "      <td>61.406863</td>\n",
       "      <td>20.410189</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>F. hepatica</td>\n",
       "      <td>2,816</td>\n",
       "      <td>1901411</td>\n",
       "      <td>9731</td>\n",
       "      <td>7.470969</td>\n",
       "      <td>18.816155</td>\n",
       "      <td>74.966602</td>\n",
       "      <td>93.120446</td>\n",
       "      <td>75.303644</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>S. mattheei</td>\n",
       "      <td>98</td>\n",
       "      <td>48524730</td>\n",
       "      <td>10770</td>\n",
       "      <td>5.561746</td>\n",
       "      <td>21.058496</td>\n",
       "      <td>77.437326</td>\n",
       "      <td>81.857425</td>\n",
       "      <td>76.569038</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>S. curassoni</td>\n",
       "      <td>370</td>\n",
       "      <td>46058712</td>\n",
       "      <td>10084</td>\n",
       "      <td>5.474018</td>\n",
       "      <td>21.410155</td>\n",
       "      <td>79.125347</td>\n",
       "      <td>82.451159</td>\n",
       "      <td>83.816014</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>E. multilocularis</td>\n",
       "      <td>1,217</td>\n",
       "      <td>13762452</td>\n",
       "      <td>10663</td>\n",
       "      <td>7.849573</td>\n",
       "      <td>20.003751</td>\n",
       "      <td>69.905280</td>\n",
       "      <td>73.631789</td>\n",
       "      <td>66.019417</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>S. mansoni</td>\n",
       "      <td>10</td>\n",
       "      <td>49793710</td>\n",
       "      <td>9896</td>\n",
       "      <td>7.134196</td>\n",
       "      <td>23.170978</td>\n",
       "      <td>75.677041</td>\n",
       "      <td>80.619629</td>\n",
       "      <td>55.276382</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>T. asiatica</td>\n",
       "      <td>6,900</td>\n",
       "      <td>342420</td>\n",
       "      <td>13322</td>\n",
       "      <td>10.996847</td>\n",
       "      <td>19.411500</td>\n",
       "      <td>58.392133</td>\n",
       "      <td>62.245263</td>\n",
       "      <td>46.919918</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>S. margrebowiei</td>\n",
       "      <td>36</td>\n",
       "      <td>48858358</td>\n",
       "      <td>9780</td>\n",
       "      <td>6.032720</td>\n",
       "      <td>22.177914</td>\n",
       "      <td>77.453988</td>\n",
       "      <td>81.759334</td>\n",
       "      <td>72.591006</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>C. sinensis</td>\n",
       "      <td>78</td>\n",
       "      <td>168711085</td>\n",
       "      <td>13489</td>\n",
       "      <td>9.719030</td>\n",
       "      <td>17.555045</td>\n",
       "      <td>58.655201</td>\n",
       "      <td>82.816520</td>\n",
       "      <td>30.639731</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>E. granulosus</td>\n",
       "      <td>1,288</td>\n",
       "      <td>5228736</td>\n",
       "      <td>10236</td>\n",
       "      <td>8.001172</td>\n",
       "      <td>20.310668</td>\n",
       "      <td>68.454474</td>\n",
       "      <td>72.714487</td>\n",
       "      <td>66.917293</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>M. corti</td>\n",
       "      <td>7,068</td>\n",
       "      <td>65816</td>\n",
       "      <td>14704</td>\n",
       "      <td>10.935800</td>\n",
       "      <td>16.274483</td>\n",
       "      <td>58.739119</td>\n",
       "      <td>72.029555</td>\n",
       "      <td>31.271145</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>S. haematobium</td>\n",
       "      <td>163</td>\n",
       "      <td>48328128</td>\n",
       "      <td>9431</td>\n",
       "      <td>5.831831</td>\n",
       "      <td>21.842859</td>\n",
       "      <td>79.228078</td>\n",
       "      <td>81.811321</td>\n",
       "      <td>58.600583</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>S. japonicum</td>\n",
       "      <td>1,789</td>\n",
       "      <td>1093989</td>\n",
       "      <td>10089</td>\n",
       "      <td>6.462484</td>\n",
       "      <td>20.864308</td>\n",
       "      <td>74.358212</td>\n",
       "      <td>83.033932</td>\n",
       "      <td>56.298201</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>E. canadensis</td>\n",
       "      <td>9,330</td>\n",
       "      <td>74230</td>\n",
       "      <td>11425</td>\n",
       "      <td>7.527352</td>\n",
       "      <td>19.921225</td>\n",
       "      <td>57.487965</td>\n",
       "      <td>70.287671</td>\n",
       "      <td>16.901408</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>S. mediterranea</td>\n",
       "      <td>43,294</td>\n",
       "      <td>40740</td>\n",
       "      <td>29850</td>\n",
       "      <td>5.453936</td>\n",
       "      <td>14.040201</td>\n",
       "      <td>71.403685</td>\n",
       "      <td>72.334755</td>\n",
       "      <td>68.054201</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>O. viverrini</td>\n",
       "      <td>42,216</td>\n",
       "      <td>1347703</td>\n",
       "      <td>16379</td>\n",
       "      <td>8.663533</td>\n",
       "      <td>13.169302</td>\n",
       "      <td>52.457415</td>\n",
       "      <td>82.467732</td>\n",
       "      <td>54.035088</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              Species Total number of scaffolds        N50  \\\n",
       "0           T. solium                    11,237      67829   \n",
       "1          M. lignano                     5,269     244885   \n",
       "2          T. regenti                       573  125840967   \n",
       "3       H. microstoma                         7   25815276   \n",
       "4         T. saginata                     3,626     585232   \n",
       "5         H. diminuta                       719    2331359   \n",
       "6         F. hepatica                     2,816    1901411   \n",
       "7         S. mattheei                        98   48524730   \n",
       "8        S. curassoni                       370   46058712   \n",
       "9   E. multilocularis                     1,217   13762452   \n",
       "10         S. mansoni                        10   49793710   \n",
       "11        T. asiatica                     6,900     342420   \n",
       "12    S. margrebowiei                        36   48858358   \n",
       "13        C. sinensis                        78  168711085   \n",
       "14      E. granulosus                     1,288    5228736   \n",
       "15           M. corti                     7,068      65816   \n",
       "16     S. haematobium                       163   48328128   \n",
       "17       S. japonicum                     1,789    1093989   \n",
       "18      E. canadensis                     9,330      74230   \n",
       "19    S. mediterranea                    43,294      40740   \n",
       "20       O. viverrini                    42,216    1347703   \n",
       "\n",
       "    Total number of proteins  Secreted protein-coding genes (%)  \\\n",
       "0                      12356                           7.389123   \n",
       "1                      49013                           8.387571   \n",
       "2                      14478                           8.281531   \n",
       "3                      10139                           6.973074   \n",
       "4                      13160                          10.471125   \n",
       "5                      15165                           8.578965   \n",
       "6                       9731                           7.470969   \n",
       "7                      10770                           5.561746   \n",
       "8                      10084                           5.474018   \n",
       "9                      10663                           7.849573   \n",
       "10                      9896                           7.134196   \n",
       "11                     13322                          10.996847   \n",
       "12                      9780                           6.032720   \n",
       "13                     13489                           9.719030   \n",
       "14                     10236                           8.001172   \n",
       "15                     14704                          10.935800   \n",
       "16                      9431                           5.831831   \n",
       "17                     10089                           6.462484   \n",
       "18                     11425                           7.527352   \n",
       "19                     29850                           5.453936   \n",
       "20                     16379                           8.663533   \n",
       "\n",
       "    Transmembrane protein-coding genes (%)  Genes with GO annotation (%)  \\\n",
       "0                                17.651343                     71.819359   \n",
       "1                                22.573603                     71.072981   \n",
       "2                                20.251416                     69.836994   \n",
       "3                                19.666634                     71.841404   \n",
       "4                                19.247720                     58.168693   \n",
       "5                                15.522585                     55.351137   \n",
       "6                                18.816155                     74.966602   \n",
       "7                                21.058496                     77.437326   \n",
       "8                                21.410155                     79.125347   \n",
       "9                                20.003751                     69.905280   \n",
       "10                               23.170978                     75.677041   \n",
       "11                               19.411500                     58.392133   \n",
       "12                               22.177914                     77.453988   \n",
       "13                               17.555045                     58.655201   \n",
       "14                               20.310668                     68.454474   \n",
       "15                               16.274483                     58.739119   \n",
       "16                               21.842859                     79.228078   \n",
       "17                               20.864308                     74.358212   \n",
       "18                               19.921225                     57.487965   \n",
       "19                               14.040201                     71.403685   \n",
       "20                               13.169302                     52.457415   \n",
       "\n",
       "    Homologs with GO annotation (%)  Inparalogs with GO annotation (%)  \\\n",
       "0                         80.299703                          74.025974   \n",
       "1                         73.827860                          73.362007   \n",
       "2                         78.081406                          62.964099   \n",
       "3                         78.522167                          55.179283   \n",
       "4                         62.506918                          42.793462   \n",
       "5                         61.406863                          20.410189   \n",
       "6                         93.120446                          75.303644   \n",
       "7                         81.857425                          76.569038   \n",
       "8                         82.451159                          83.816014   \n",
       "9                         73.631789                          66.019417   \n",
       "10                        80.619629                          55.276382   \n",
       "11                        62.245263                          46.919918   \n",
       "12                        81.759334                          72.591006   \n",
       "13                        82.816520                          30.639731   \n",
       "14                        72.714487                          66.917293   \n",
       "15                        72.029555                          31.271145   \n",
       "16                        81.811321                          58.600583   \n",
       "17                        83.033932                          56.298201   \n",
       "18                        70.287671                          16.901408   \n",
       "19                        72.334755                          68.054201   \n",
       "20                        82.467732                          54.035088   \n",
       "\n",
       "    Unannotated inparalogs (%)  \n",
       "0                          0.0  \n",
       "1                          0.0  \n",
       "2                          0.0  \n",
       "3                          0.0  \n",
       "4                          0.0  \n",
       "5                          0.0  \n",
       "6                          0.0  \n",
       "7                          0.0  \n",
       "8                          0.0  \n",
       "9                          0.0  \n",
       "10                         0.0  \n",
       "11                         0.0  \n",
       "12                         0.0  \n",
       "13                         0.0  \n",
       "14                         0.0  \n",
       "15                         0.0  \n",
       "16                         0.0  \n",
       "17                         0.0  \n",
       "18                         0.0  \n",
       "19                         0.0  \n",
       "20                         0.0  "
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "supp_table_two"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplementary Table 3: GO terms enrichment analysis results\n",
    "This table has already being produced at the **functional_enrichment_analysis.ipynb** notebook."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplementary Table 4: \n",
    "The raw table for this Supp. Table was created in the **molecular_evolution_analyses.ipynb** notebook, as well as in the notebook defining inparalogs and computing InterPro annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load raw table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "codeml_results = pd.read_csv('../results/molecular_evolution_analyses/codeml_results/codeml_results.tsv', sep ='\\t').drop('index', axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = codeml_results\n",
    "# Create a dictionary for LnL values when Model is M1\n",
    "lnl_dict = df[df['Model'] == 'M1'].set_index('Family')['LnL'].to_dict()\n",
    "\n",
    "# Subset the DataFrame to keep only Model M2 and rename the LnL column to LnL_2\n",
    "m2_df = df[df['Model'] == 'M2'].copy()\n",
    "m2_df.rename(columns={'LnL': 'LnL_2'}, inplace=True)\n",
    "\n",
    "# Assign a new column for LnL_1 using the dictionary\n",
    "m2_df['LnL_1'] = m2_df['Family'].map(lnl_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "supp_table_4 = m2_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add Genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "mgc2genes = {row['monophyletic_group_code']: row['Genes'] for index,row in pd.read_csv('../results/misc/inparalogs_group_composition_final.tsv', sep = '\\t').iterrows()}\n",
    "mgc2species = {row['monophyletic_group_code']: row['Species'] for index,row in pd.read_csv('../results/misc/inparalogs_group_composition_final.tsv', sep = '\\t').iterrows()}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "# add Genes and Species\n",
    "supp_table_4 = supp_table_4.assign(Genes = lambda df: [mgc2genes[row['Family']] for index,row in df.iterrows()],\n",
    "                                   Species = lambda df: [mgc2species[row['Family']] for index,row in df.iterrows()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "supp_table_4 = supp_table_4[['Family', 'Genes', 'Species', 'LnL_1', 'LnL_2', 'Likelihood_ratio', 'Num. sites (BEB pp > 0.95)', 'Num. sites (BEB pp > 0.99)',\n",
    "              'dN', 'dS', 'dN/dS', 'Significativo']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "# set table in correct order"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Family</th>\n",
       "      <th>Genes</th>\n",
       "      <th>Species</th>\n",
       "      <th>LnL_1</th>\n",
       "      <th>LnL_2</th>\n",
       "      <th>Likelihood_ratio</th>\n",
       "      <th>Num. sites (BEB pp &gt; 0.95)</th>\n",
       "      <th>Num. sites (BEB pp &gt; 0.99)</th>\n",
       "      <th>dN</th>\n",
       "      <th>dS</th>\n",
       "      <th>dN/dS</th>\n",
       "      <th>Significativo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>F79292_TASI_G1</td>\n",
       "      <td>TASs00700g11196m00001, TASs00797g11394m00001, ...</td>\n",
       "      <td>T. asiatica</td>\n",
       "      <td>-561.882472</td>\n",
       "      <td>-561.882470</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.5359</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>F8315_TASI_G1</td>\n",
       "      <td>TASs00257g09507m00001, TASs00257g09505m00001</td>\n",
       "      <td>T. asiatica</td>\n",
       "      <td>-75.424929</td>\n",
       "      <td>-75.424872</td>\n",
       "      <td>0.000114</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>F31918_MLIG_G1</td>\n",
       "      <td>BOX15_Mlig003887g1, BOX15_Mlig003887g3, BOX15_...</td>\n",
       "      <td>M. lignano</td>\n",
       "      <td>-1319.672893</td>\n",
       "      <td>-1319.539216</td>\n",
       "      <td>0.267354</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.7685</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>F23474_MLIG_G1</td>\n",
       "      <td>BOX15_Mlig006866g1, BOX15_Mlig006866g2, BOX15_...</td>\n",
       "      <td>M. lignano</td>\n",
       "      <td>-874.963685</td>\n",
       "      <td>-874.963684</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.3769</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>F53749_MLIG_G1</td>\n",
       "      <td>BOX15_Mlig031608g3, BOX15_Mlig031608g2, BOX15_...</td>\n",
       "      <td>M. lignano</td>\n",
       "      <td>-1140.362246</td>\n",
       "      <td>-1140.362241</td>\n",
       "      <td>0.000010</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0058</td>\n",
       "      <td>0.0259</td>\n",
       "      <td>0.2219</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39782</th>\n",
       "      <td>F129217_MLIG_G1</td>\n",
       "      <td>BOX15_Mlig022362g1, BOX15_Mlig026567g1, BOX15_...</td>\n",
       "      <td>M. lignano</td>\n",
       "      <td>-1717.806086</td>\n",
       "      <td>-1711.561286</td>\n",
       "      <td>12.489600</td>\n",
       "      <td>74.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>9.3191</td>\n",
       "      <td>0.7274</td>\n",
       "      <td>12.8117</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39784</th>\n",
       "      <td>F93693_MLIG_G1</td>\n",
       "      <td>BOX15_Mlig018049g7, BOX15_Mlig018049g1</td>\n",
       "      <td>M. lignano</td>\n",
       "      <td>-1966.452456</td>\n",
       "      <td>-1966.014225</td>\n",
       "      <td>0.876462</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>124.1390</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39787</th>\n",
       "      <td>F114209_MLIG_G1</td>\n",
       "      <td>BOX15_Mlig011283g1, BOX15_Mlig011283g6, BOX15_...</td>\n",
       "      <td>M. lignano</td>\n",
       "      <td>-213.669497</td>\n",
       "      <td>-213.669409</td>\n",
       "      <td>0.000176</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0734</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39789</th>\n",
       "      <td>F80846_MLIG_G1</td>\n",
       "      <td>BOX15_Mlig033116g1, BOX15_Mlig033116g2, BOX15_...</td>\n",
       "      <td>M. lignano</td>\n",
       "      <td>-1722.066589</td>\n",
       "      <td>-1718.970376</td>\n",
       "      <td>6.192426</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0241</td>\n",
       "      <td>0.0079</td>\n",
       "      <td>3.0504</td>\n",
       "      <td>***</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>39790</th>\n",
       "      <td>F86787_TREG_G1</td>\n",
       "      <td>TREG1_105450.1, TREG1_117310.1</td>\n",
       "      <td>T. regenti</td>\n",
       "      <td>-697.109518</td>\n",
       "      <td>-697.109505</td>\n",
       "      <td>0.000026</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.1047</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>19896 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                Family                                              Genes  \\\n",
       "0       F79292_TASI_G1  TASs00700g11196m00001, TASs00797g11394m00001, ...   \n",
       "3        F8315_TASI_G1       TASs00257g09507m00001, TASs00257g09505m00001   \n",
       "4       F31918_MLIG_G1  BOX15_Mlig003887g1, BOX15_Mlig003887g3, BOX15_...   \n",
       "6       F23474_MLIG_G1  BOX15_Mlig006866g1, BOX15_Mlig006866g2, BOX15_...   \n",
       "8       F53749_MLIG_G1  BOX15_Mlig031608g3, BOX15_Mlig031608g2, BOX15_...   \n",
       "...                ...                                                ...   \n",
       "39782  F129217_MLIG_G1  BOX15_Mlig022362g1, BOX15_Mlig026567g1, BOX15_...   \n",
       "39784   F93693_MLIG_G1             BOX15_Mlig018049g7, BOX15_Mlig018049g1   \n",
       "39787  F114209_MLIG_G1  BOX15_Mlig011283g1, BOX15_Mlig011283g6, BOX15_...   \n",
       "39789   F80846_MLIG_G1  BOX15_Mlig033116g1, BOX15_Mlig033116g2, BOX15_...   \n",
       "39790   F86787_TREG_G1                     TREG1_105450.1, TREG1_117310.1   \n",
       "\n",
       "           Species        LnL_1        LnL_2  Likelihood_ratio  \\\n",
       "0      T. asiatica  -561.882472  -561.882470          0.000004   \n",
       "3      T. asiatica   -75.424929   -75.424872          0.000114   \n",
       "4       M. lignano -1319.672893 -1319.539216          0.267354   \n",
       "6       M. lignano  -874.963685  -874.963684          0.000002   \n",
       "8       M. lignano -1140.362246 -1140.362241          0.000010   \n",
       "...            ...          ...          ...               ...   \n",
       "39782   M. lignano -1717.806086 -1711.561286         12.489600   \n",
       "39784   M. lignano -1966.452456 -1966.014225          0.876462   \n",
       "39787   M. lignano  -213.669497  -213.669409          0.000176   \n",
       "39789   M. lignano -1722.066589 -1718.970376          6.192426   \n",
       "39790   T. regenti  -697.109518  -697.109505          0.000026   \n",
       "\n",
       "       Num. sites (BEB pp > 0.95)  Num. sites (BEB pp > 0.99)      dN      dS  \\\n",
       "0                             0.0                         0.0  0.0000  0.0000   \n",
       "3                             0.0                         0.0  0.0000  0.0000   \n",
       "4                             0.0                         0.0  0.0000  0.0000   \n",
       "6                             0.0                         0.0  0.0000  0.0000   \n",
       "8                             0.0                         0.0  0.0058  0.0259   \n",
       "...                           ...                         ...     ...     ...   \n",
       "39782                        74.0                        34.0  9.3191  0.7274   \n",
       "39784                         0.0                         0.0  0.0000  0.0000   \n",
       "39787                         0.0                         0.0  0.0000  0.0734   \n",
       "39789                         0.0                         0.0  0.0241  0.0079   \n",
       "39790                         0.0                         0.0  0.0000  0.0000   \n",
       "\n",
       "          dN/dS Significativo  \n",
       "0        0.5359             -  \n",
       "3        0.0000             -  \n",
       "4        0.7685             -  \n",
       "6        0.3769             -  \n",
       "8        0.2219             -  \n",
       "...         ...           ...  \n",
       "39782   12.8117           ***  \n",
       "39784  124.1390             -  \n",
       "39787    0.0000             -  \n",
       "39789    3.0504           ***  \n",
       "39790    0.1047             -  \n",
       "\n",
       "[19896 rows x 12 columns]"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "supp_table_4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to get InterPro annotation for a gene\n",
    "def get_interpro_annotation(gene):\n",
    "    return code2interpro.get(gene, \"\")\n",
    "\n",
    "# Apply the function to each row to get InterPro annotations\n",
    "supp_table_4['InterPro Annotation'] = supp_table_4['Genes'].apply(lambda x: '\\n'.join([get_interpro_annotation(g) for g in x.split(', ')]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [],
   "source": [
    "# rename and save tsv\n",
    "supp_table_4.rename({'Family': 'Inparalogs cluster code #DB',\n",
    "                     'LnL_1': 'lnL (M1a = nearly neutral)',\n",
    "                     'LnL_2': 'lnL (M2a = selection allowed)',\n",
    "                     'dN/dS': 'Global dN/dS',\n",
    "                     'Likelihood_ratio': 'LRT',\n",
    "                     'Significativo': 'Significance (LRT > 6)',\n",
    "                     'Num. sites (BEB pp > 0.95)': '# sites (BEB pp > 0.95)',\n",
    "                     'Num. sites (BEB pp > 0.95)': '# sites (BEB pp > 0.95)', \n",
    "                     'InterPro Annotation': 'InterPro Annotations'}, axis = 1).to_csv('../results/tables/supplementary_table_4.tsv', sep = '\\t', index = False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplementary Table 5\n",
    "This table has already being produced (see notebooks/scripts in this folder)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Supplementary Table 6\n",
    "This table has already being produced at the **renaming_set_filtering_and_homology_inferences.ipynb** notebook."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
